// deck.gl/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.experimental = exports.MapController = exports.ViewportController = exports.DeckGL = exports.default = exports.ExtrudedChoroplethLayer64 = exports.ChoroplethLayer64 = exports.ChoroplethLayer = exports.GeoJsonLayer = exports.PolygonLayer = exports.PathLayer = exports.HexagonCellLayer = exports.HexagonLayer = exports.GridCellLayer = exports.GridLayer = exports.ScreenGridLayer = exports.ScatterplotLayer = exports.PointCloudLayer = exports.LineLayer = exports.IconLayer = exports.ArcLayer = exports.assembleShaders = exports.OrthographicViewport = exports.PerspectiveViewport = exports.MapState = exports.OrbitState = exports.FirstPersonState = exports.ViewportControls = exports.WebMercatorViewport = exports.ThirdPersonViewport = exports.FirstPersonViewport = exports.Viewport = exports.CompositeLayer = exports.Layer = exports.AttributeManager = exports.LayerManager = exports.COORDINATE_SYSTEM = undefined;

var _constants = require('./core/lib/constants');

Object.defineProperty(exports, 'COORDINATE_SYSTEM', {
  enumerable: true,
  get: function get() {
    return _constants.COORDINATE_SYSTEM;
  }
});

var _layerManager = require('./core/lib/layer-manager');

Object.defineProperty(exports, 'LayerManager', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_layerManager).default;
  }
});

var _attributeManager = require('./core/lib/attribute-manager');

Object.defineProperty(exports, 'AttributeManager', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_attributeManager).default;
  }
});

var _layer = require('./core/lib/layer');

Object.defineProperty(exports, 'Layer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_layer).default;
  }
});

var _compositeLayer = require('./core/lib/composite-layer');

Object.defineProperty(exports, 'CompositeLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_compositeLayer).default;
  }
});

var _viewport = require('./core/viewports/viewport');

Object.defineProperty(exports, 'Viewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_viewport).default;
  }
});

var _firstPersonViewport = require('./core/viewports/first-person-viewport');

Object.defineProperty(exports, 'FirstPersonViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_firstPersonViewport).default;
  }
});

var _thirdPersonViewport = require('./core/viewports/third-person-viewport');

Object.defineProperty(exports, 'ThirdPersonViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_thirdPersonViewport).default;
  }
});

var _webMercatorViewport = require('./core/viewports/web-mercator-viewport');

Object.defineProperty(exports, 'WebMercatorViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webMercatorViewport).default;
  }
});

var _controls = require('./core/controllers/controls');

Object.defineProperty(exports, 'ViewportControls', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_controls).default;
  }
});

var _firstPersonState = require('./core/controllers/first-person-state');

Object.defineProperty(exports, 'FirstPersonState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_firstPersonState).default;
  }
});

var _orbitState = require('./core/controllers/orbit-state');

Object.defineProperty(exports, 'OrbitState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_orbitState).default;
  }
});

var _mapState = require('./core/controllers/map-state');

Object.defineProperty(exports, 'MapState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mapState).default;
  }
});

var _perspectiveViewport = require('./core/deprecated/viewports/perspective-viewport');

Object.defineProperty(exports, 'PerspectiveViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_perspectiveViewport).default;
  }
});

var _orthographicViewport = require('./core/deprecated/viewports/orthographic-viewport');

Object.defineProperty(exports, 'OrthographicViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_orthographicViewport).default;
  }
});

var _luma = require('luma.gl');

Object.defineProperty(exports, 'assembleShaders', {
  enumerable: true,
  get: function get() {
    return _luma.assembleShaders;
  }
});

var _arcLayer = require('./core-layers/arc-layer/arc-layer');

Object.defineProperty(exports, 'ArcLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_arcLayer).default;
  }
});

var _iconLayer = require('./core-layers/icon-layer/icon-layer');

Object.defineProperty(exports, 'IconLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_iconLayer).default;
  }
});

var _lineLayer = require('./core-layers/line-layer/line-layer');

Object.defineProperty(exports, 'LineLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lineLayer).default;
  }
});

var _pointCloudLayer = require('./core-layers/point-cloud-layer/point-cloud-layer');

Object.defineProperty(exports, 'PointCloudLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pointCloudLayer).default;
  }
});

var _scatterplotLayer = require('./core-layers/scatterplot-layer/scatterplot-layer');

Object.defineProperty(exports, 'ScatterplotLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scatterplotLayer).default;
  }
});

var _screenGridLayer = require('./core-layers/screen-grid-layer/screen-grid-layer');

Object.defineProperty(exports, 'ScreenGridLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_screenGridLayer).default;
  }
});

var _gridLayer = require('./core-layers/grid-layer/grid-layer');

Object.defineProperty(exports, 'GridLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gridLayer).default;
  }
});

var _gridCellLayer = require('./core-layers/grid-cell-layer/grid-cell-layer');

Object.defineProperty(exports, 'GridCellLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gridCellLayer).default;
  }
});

var _hexagonLayer = require('./core-layers/hexagon-layer/hexagon-layer');

Object.defineProperty(exports, 'HexagonLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hexagonLayer).default;
  }
});

var _hexagonCellLayer = require('./core-layers/hexagon-cell-layer/hexagon-cell-layer');

Object.defineProperty(exports, 'HexagonCellLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hexagonCellLayer).default;
  }
});

var _pathLayer = require('./core-layers/path-layer/path-layer');

Object.defineProperty(exports, 'PathLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pathLayer).default;
  }
});

var _polygonLayer = require('./core-layers/polygon-layer/polygon-layer');

Object.defineProperty(exports, 'PolygonLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_polygonLayer).default;
  }
});

var _geojsonLayer = require('./core-layers/geojson-layer/geojson-layer');

Object.defineProperty(exports, 'GeoJsonLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_geojsonLayer).default;
  }
});

var _choroplethLayer = require('./deprecated-layers/choropleth-layer/choropleth-layer');

Object.defineProperty(exports, 'ChoroplethLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_choroplethLayer).default;
  }
});

var _choroplethLayer2 = require('./deprecated-layers/choropleth-layer-64/choropleth-layer-64');

Object.defineProperty(exports, 'ChoroplethLayer64', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_choroplethLayer2).default;
  }
});

var _extrudedChoroplethLayer = require('./deprecated-layers/extruded-choropleth-layer-64/extruded-choropleth-layer-64');

Object.defineProperty(exports, 'ExtrudedChoroplethLayer64', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_extrudedChoroplethLayer).default;
  }
});

var _deckgl = require('./react/deckgl');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_deckgl).default;
  }
});
Object.defineProperty(exports, 'DeckGL', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_deckgl).default;
  }
});

var _viewportController = require('./react/viewport-controller');

Object.defineProperty(exports, 'ViewportController', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_viewportController).default;
  }
});

var _mapController = require('./react/map-controller');

Object.defineProperty(exports, 'MapController', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mapController).default;
  }
});

var _deck = require('./core/pure-js/deck');

var _deck2 = _interopRequireDefault(_deck);

var _mapController2 = require('./core/pure-js/map-controller');

var _mapController3 = _interopRequireDefault(_mapController2);

var _get = require('./core/lib/utils/get');

var _count = require('./core/lib/utils/count');

var _effectManager = require('./core/experimental/lib/effect-manager');

var _effectManager2 = _interopRequireDefault(_effectManager);

var _effect = require('./core/experimental/lib/effect');

var _effect2 = _interopRequireDefault(_effect);

var _reflectionEffect = require('./effects/experimental/reflection-effect/reflection-effect');

var _reflectionEffect2 = _interopRequireDefault(_reflectionEffect);

var _orbitController = require('./react/experimental/orbit-controller');

var _orbitController2 = _interopRequireDefault(_orbitController);

var _animationMapController = require('./react/experimental/animation-map-controller');

var _animationMapController2 = _interopRequireDefault(_animationMapController);

var _viewportAnimationUtils = require('./react/experimental/viewport-animation-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
/* eslint-disable max-len */

var experimental = {};

//
// CORE LIBRARY
//

// Experimental Pure JS (non-React) bindings


// Experimental Features (May change in minor version bumps, use at your own risk)
// Forward the luma.gl version (note: now integrated with Model)

Object.assign(experimental, {
  get: _get.get,
  count: _count.count,
  EffectManager: _effectManager2.default,
  Effect: _effect2.default,

  // Pure JS (i.e. non-React) support
  DeckGLJS: _deck2.default, // Integrate into `LayerManager`?
  MapControllerJS: _mapController3.default // Integrate into `MapController` JS class?
});

//
// CORE LAYERS PACKAGE
//

//
// EFFECTS PACKAGE
//

Object.assign(experimental, {
  ReflectionEffect: _reflectionEffect2.default
});

//
// REACT BINDINGS PACKAGE
//

// Experimental React bindings


// Experimental react bindings
Object.assign(experimental, {
  OrbitController: _orbitController2.default,
  AnimationMapController: _animationMapController2.default,
  viewportLinearInterpolator: _viewportAnimationUtils.viewportLinearInterpolator,
  viewportFlyToInterpolator: _viewportAnimationUtils.viewportFlyToInterpolator,
  ANIMATION_EVENTS: _animationMapController.ANIMATION_EVENTS
});

//
// EXPERIMENTAL EXPORTS
//

exports.experimental = experimental;
