// deck.gl/dist/core/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.experimental = exports.MapState = exports.OrbitState = exports.FirstPersonState = exports.ViewportControls = exports.WebMercatorViewport = exports.ThirdPersonViewport = exports.FirstPersonViewport = exports.Viewport = exports.CompositeLayer = exports.Layer = exports.AttributeManager = exports.LayerManager = exports.COORDINATE_SYSTEM = undefined;

var _constants = require('./lib/constants');

Object.defineProperty(exports, 'COORDINATE_SYSTEM', {
  enumerable: true,
  get: function get() {
    return _constants.COORDINATE_SYSTEM;
  }
});

var _layerManager = require('./lib/layer-manager');

Object.defineProperty(exports, 'LayerManager', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_layerManager).default;
  }
});

var _attributeManager = require('./lib/attribute-manager');

Object.defineProperty(exports, 'AttributeManager', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_attributeManager).default;
  }
});

var _layer = require('./lib/layer');

Object.defineProperty(exports, 'Layer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_layer).default;
  }
});

var _compositeLayer = require('./lib/composite-layer');

Object.defineProperty(exports, 'CompositeLayer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_compositeLayer).default;
  }
});

var _viewport = require('./viewports/viewport');

Object.defineProperty(exports, 'Viewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_viewport).default;
  }
});

var _firstPersonViewport = require('./viewports/first-person-viewport');

Object.defineProperty(exports, 'FirstPersonViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_firstPersonViewport).default;
  }
});

var _thirdPersonViewport = require('./viewports/third-person-viewport');

Object.defineProperty(exports, 'ThirdPersonViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_thirdPersonViewport).default;
  }
});

var _webMercatorViewport = require('./viewports/web-mercator-viewport');

Object.defineProperty(exports, 'WebMercatorViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webMercatorViewport).default;
  }
});

var _controls = require('./controllers/controls');

Object.defineProperty(exports, 'ViewportControls', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_controls).default;
  }
});

var _firstPersonState = require('./controllers/first-person-state');

Object.defineProperty(exports, 'FirstPersonState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_firstPersonState).default;
  }
});

var _orbitState = require('./controllers/orbit-state');

Object.defineProperty(exports, 'OrbitState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_orbitState).default;
  }
});

var _mapState = require('./controllers/map-state');

Object.defineProperty(exports, 'MapState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mapState).default;
  }
});

var _deck = require('./pure-js/deck');

var _deck2 = _interopRequireDefault(_deck);

var _mapController = require('./pure-js/map-controller');

var _mapController2 = _interopRequireDefault(_mapController);

var _get = require('./lib/utils/get');

var _count = require('./lib/utils/count');

var _effectManager = require('./experimental/lib/effect-manager');

var _effectManager2 = _interopRequireDefault(_effectManager);

var _effect = require('./experimental/lib/effect');

var _effect2 = _interopRequireDefault(_effect);

var _binSorter = require('./utils/bin-sorter');

var _binSorter2 = _interopRequireDefault(_binSorter);

var _scaleUtils = require('./utils/scale-utils');

var _colorUtils = require('./utils/color-utils');

var _fp = require('./lib/utils/fp64');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Experimental Pure JS (non-React) bindings


// Experimental Features (May change in minor version bumps, use at your own risk)


// TODO - just expose as layer methods instead?


var experimental = exports.experimental = {
  BinSorter: _binSorter2.default,
  linearScale: _scaleUtils.linearScale,
  quantizeScale: _scaleUtils.quantizeScale,
  clamp: _scaleUtils.clamp,
  defaultColorRange: _colorUtils.defaultColorRange,

  get: _get.get,
  count: _count.count,

  enable64bitSupport: _fp.enable64bitSupport,
  fp64ify: _fp.fp64ify,

  EffectManager: _effectManager2.default,
  Effect: _effect2.default,

  DeckGLJS: _deck2.default,
  MapControllerJS: _mapController2.default
};
