// deck.gl/dist/core/lib/utils/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _flatten = require('./flatten');

Object.defineProperty(exports, 'flatten', {
  enumerable: true,
  get: function get() {
    return _flatten.flatten;
  }
});
Object.defineProperty(exports, 'flattenVertices', {
  enumerable: true,
  get: function get() {
    return _flatten.flattenVertices;
  }
});
Object.defineProperty(exports, 'fillArray', {
  enumerable: true,
  get: function get() {
    return _flatten.fillArray;
  }
});

var _compareObjects = require('./compare-objects');

Object.defineProperty(exports, 'compareProps', {
  enumerable: true,
  get: function get() {
    return _compareObjects.compareProps;
  }
});
Object.defineProperty(exports, 'areEqualShallow', {
  enumerable: true,
  get: function get() {
    return _compareObjects.areEqualShallow;
  }
});

var _compareArrays = require('./compare-arrays');

Object.defineProperty(exports, 'compareArrays', {
  enumerable: true,
  get: function get() {
    return _compareArrays.compareArrays;
  }
});
Object.defineProperty(exports, 'checkArray', {
  enumerable: true,
  get: function get() {
    return _compareArrays.checkArray;
  }
});

var _fp = require('./fp64');

Object.defineProperty(exports, 'fp64ify', {
  enumerable: true,
  get: function get() {
    return _fp.fp64ify;
  }
});

var _color = require('./color');

Object.defineProperty(exports, 'parseColor', {
  enumerable: true,
  get: function get() {
    return _color.parseColor;
  }
});

var _get = require('./get');

Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function get() {
    return _get.get;
  }
});

var _count = require('./count');

Object.defineProperty(exports, 'count', {
  enumerable: true,
  get: function get() {
    return _count.count;
  }
});

var _log = require('../../utils/log');

Object.defineProperty(exports, 'log', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_log).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
